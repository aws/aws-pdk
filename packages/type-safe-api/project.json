{
  "name": "@aws/type-safe-api",
  "root": "packages/type-safe-api",
  "targets": {
    "build": {
      "inputs": [
        "default",
        "^default",
        "!{projectRoot}/.jsii",
        "!{projectRoot}/lib/**/*",
        "!{projectRoot}/dist/**/*",
        "!{projectRoot}/coverage/**/*",
        "!{projectRoot}/test-reports/**/*",
        "!{projectRoot}/docs/api/**/*"
      ],
      "outputs": [
        "{projectRoot}/.jsii",
        "{projectRoot}/lib",
        "{projectRoot}/dist",
        "{projectRoot}/coverage",
        "{projectRoot}/test-reports",
        "{projectRoot}/docs/api"
      ],
      "dependsOn": [
        "^build"
      ],
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen build",
        "cwd": "packages/type-safe-api"
      }
    },
    "default": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen default",
        "cwd": "packages/type-safe-api"
      }
    },
    "pre-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen pre-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "post-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "test": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test",
        "cwd": "packages/type-safe-api"
      }
    },
    "package": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package",
        "cwd": "packages/type-safe-api"
      }
    },
    "test:watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test:watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint",
        "cwd": "packages/type-safe-api"
      }
    },
    "compat": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compat",
        "cwd": "packages/type-safe-api"
      }
    },
    "package-all": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package-all",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:java": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:java",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:python": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:python",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint-staged": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint-staged",
        "cwd": "packages/type-safe-api"
      }
    },
    "jest": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen jest",
        "cwd": "packages/type-safe-api"
      }
    },
    "docgen": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen docgen",
        "cwd": "packages/type-safe-api"
      }
    }
  },
  "implicitDependencies": [
    "smithy-async-transformer"
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"pnpm exec projen\"."
}
