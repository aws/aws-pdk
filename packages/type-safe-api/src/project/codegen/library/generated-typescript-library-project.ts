/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { NodePackageUtils } from "@aws/monorepo";
import { NodePackageManager } from "projen/lib/javascript";
import {
  TypeScriptProject,
  TypeScriptProjectOptions,
} from "projen/lib/typescript";
import {
  CodeGenerationSourceOptions,
  GeneratedWithOpenApiGeneratorOptions,
} from "../../types";
import { OpenApiGeneratorIgnoreFile } from "../components/open-api-generator-ignore-file";
import { OpenApiToolsJsonFile } from "../components/open-api-tools-json-file";
import { TypeSafeApiCommandEnvironment } from "../components/type-safe-api-command-environment";
import {
  buildCleanOpenApiGeneratedCodeCommand,
  buildInvokeOpenApiGeneratorCommandArgs,
  buildTypeSafeApiExecCommand,
  GenerationOptions,
  TypeSafeApiScript,
} from "../components/utils";

/**
 * Configuration for a generated typescript library
 */
export interface GeneratedTypescriptLibraryProjectOptions
  extends TypeScriptProjectOptions,
    GeneratedWithOpenApiGeneratorOptions,
    CodeGenerationSourceOptions {
  /**
   * Whether this project is parented by an monorepo or not
   */
  readonly isWithinMonorepo?: boolean;
}

/**
 * Typescript generated library project
 */
export abstract class GeneratedTypescriptLibraryProject extends TypeScriptProject {
  /**
   * Patterns that are excluded from code generation
   */
  public static openApiIgnorePatterns: string[] = [
    "package.json",
    "tsconfig.json",
    "tsconfig.esm.json",
    ".npmignore",
  ];

  /**
   * Options configured for the project
   */
  protected readonly options: GeneratedTypescriptLibraryProjectOptions;

  protected readonly openapiGeneratorIgnore: OpenApiGeneratorIgnoreFile;

  constructor(options: GeneratedTypescriptLibraryProjectOptions) {
    super({
      ...options,
      sampleCode: false,
      tsconfig: {
        ...options.tsconfig,
        compilerOptions: {
          lib: ["dom", "es2019"],
          // Generated code isn't very strict!
          strict: false,
          alwaysStrict: false,
          noImplicitAny: false,
          noImplicitReturns: false,
          noImplicitThis: false,
          noUnusedLocals: false,
          noUnusedParameters: false,
          strictNullChecks: false,
          strictPropertyInitialization: false,
          skipLibCheck: true,
          ...options?.tsconfig?.compilerOptions,
        },
      },
      eslint: false,
      // Disable tests unless explicitly enabled
      jest: options.jest ?? false,
      npmignoreEnabled: false,
    });
    TypeSafeApiCommandEnvironment.ensure(this);

    this.options = options;

    // Disable strict peer dependencies for pnpm as the default typescript project dependencies have type mismatches
    // (ts-jest@27 and @types/jest@28)
    if (this.package.packageManager === NodePackageManager.PNPM) {
      this.npmrc.addConfig("strict-peer-dependencies", "false");
    }

    // Tell OpenAPI Generator CLI not to generate files that we will generate via this project, or don't need.
    this.openapiGeneratorIgnore = new OpenApiGeneratorIgnoreFile(this);
    this.openapiGeneratorIgnore.addPatterns(
      ...GeneratedTypescriptLibraryProject.openApiIgnorePatterns
    );

    // Add OpenAPI Generator cli configuration
    OpenApiToolsJsonFile.ensure(this).addOpenApiGeneratorCliConfig(
      options.openApiGeneratorCliConfig
    );

    const generateTask = this.addTask("generate");
    generateTask.exec(buildCleanOpenApiGeneratedCodeCommand());
    generateTask.exec(
      buildTypeSafeApiExecCommand(
        TypeSafeApiScript.GENERATE,
        this.buildGenerateCommandArgs()
      )
    );

    this.preCompileTask.spawn(generateTask);

    if (!options.commitGeneratedCode) {
      // Ignore all the generated code
      this.gitignore.addPatterns(
        "src",
        ".npmignore",
        "README.md",
        ".openapi-generator"
      );
    }

    // If we're not in a monorepo, we need to link the generated client such that any local dependency on it can be
    // resolved
    if (!options.isWithinMonorepo) {
      switch (this.package.packageManager) {
        case NodePackageManager.PNPM:
          // Nothing to do for pnpm, since the pnpm link command handles both the dependant and dependee
          break;
        default:
          this.tasks
            .tryFind("install")
            ?.exec(
              NodePackageUtils.command.cmd(this.package.packageManager, "link")
            );
          break;
      }
    }
  }

  public buildGenerateCommandArgs = () => {
    return buildInvokeOpenApiGeneratorCommandArgs(
      this.buildOpenApiGeneratorOptions()
    );
  };

  protected abstract buildOpenApiGeneratorOptions(): GenerationOptions;
}
