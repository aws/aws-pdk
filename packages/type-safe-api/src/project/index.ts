/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./model/type-safe-api-model-build";
export * from "./model/type-safe-api-async-model-build";
export * from "./model/smithy/smithy-project-definition";
export * from "./model/openapi/open-api-project-definition";
export * from "./type-safe-api-project";
export * from "./model/type-safe-api-model-project";
export * from "./model/smithy/smithy-definition";
export * from "./model/openapi/open-api-definition";
export * from "./type-safe-websocket-api-project";
export * from "./model/type-safe-websocket-api-model-project";
export * from "./model/smithy/smithy-async-definition";
export * from "./model/openapi/open-api-async-definition";
export * from "./model/smithy/types";
export * from "./model/openapi/open-api-model-project";
export * from "./model/openapi/open-api-async-model-project";
export * from "./model/smithy/smithy-async-model-project";
export * from "./model/smithy/smithy-model-project";
export * from "./languages";
export * from "./types";
export * from "./typescript-project-options";
export * from "./python-project-options";
export * from "./java-project-options";
