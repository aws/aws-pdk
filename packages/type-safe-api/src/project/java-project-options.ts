// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import { github, GitOptions, IgnoreFileOptions, java, LoggerOptions, Project, ProjectType, ProjenrcJsonOptions, RenovatebotOptions, SampleReadmeProps } from 'projen';

/**
 * JavaProjectOptions
 */
export interface JavaProjectOptions {
  /**
   * The java package to use for the code sample.
   * @default "org.acme"
   * @stability experimental
   */
  readonly sampleJavaPackage?: string;
  /**
   * Include sample code and test if the relevant directories don't exist.
   * @default true
   * @stability experimental
   */
  readonly sample?: boolean;
  /**
   * List of test dependencies for this project.
   * Dependencies use the format: `<groupId>/<artifactId>@<semver>`
   *
   * Additional dependencies can be added via `project.addTestDependency()`.
   * @default []
   * @stability experimental
   * @featured true
   */
  readonly testDeps?: Array<string>;
  /**
   * Options related to projenrc in java.
   * @default - default options
   * @stability experimental
   */
  readonly projenrcJavaOptions?: java.ProjenrcOptions;
  /**
   * Use projenrc in java.
   * This will install `projen` as a java dependency and will add a `synth` task which
   * will compile & execute `main()` from `src/main/java/projenrc.java`.
   * @default true
   * @stability experimental
   */
  readonly projenrcJava?: boolean;
  /**
   * Packaging options.
   * @default - defaults
   * @stability experimental
   */
  readonly packagingOptions?: java.MavenPackagingOptions;
  /**
   * junit options.
   * @default - defaults
   * @stability experimental
   */
  readonly junitOptions?: java.JunitOptions;
  /**
   * Include junit tests.
   * @default true
   * @stability experimental
   */
  readonly junit?: boolean;
  /**
   * Final artifact output directory.
   * @default "dist/java"
   * @stability experimental
   */
  readonly distdir?: string;
  /**
   * List of runtime dependencies for this project.
   * Dependencies use the format: `<groupId>/<artifactId>@<semver>`
   *
   * Additional dependencies can be added via `project.addDependency()`.
   * @default []
   * @stability experimental
   * @featured true
   */
  readonly deps?: Array<string>;
  /**
   * Compile options.
   * @default - defaults
   * @stability experimental
   */
  readonly compileOptions?: java.MavenCompileOptions;
  /**
   * The URL, like the name, is not required.
   * This is a nice gesture for
   * projects users, however, so that they know where the project lives.
   * @default undefined
   * @stability experimental
   * @featured true
   */
  readonly url?: string;
  /**
   * Project packaging format.
   * @default "jar"
   * @stability experimental
   */
  readonly packaging?: string;
  /**
   * Description of a project is always good.
   * Although this should not replace
   * formal documentation, a quick comment to any readers of the POM is always
   * helpful.
   * @default undefined
   * @stability experimental
   * @featured true
   */
  readonly description?: string;
  /**
   * This is the last piece of the naming puzzle.
   * groupId:artifactId denotes a
   * single project but they cannot delineate which incarnation of that project
   * we are talking about. Do we want the junit:junit of 2018 (version 4.12), or
   * of 2007 (version 3.8.2)? In short: code changes, those changes should be
   * versioned, and this element keeps those versions in line. It is also used
   * within an artifact's repository to separate versions from each other.
   * my-project version 1.0 files live in the directory structure
   * $M2_REPO/org/codehaus/mojo/my-project/1.0.
   * @default "0.1.0"
   * @stability experimental
   * @featured true
   */
  readonly version?: string;
  /**
   * This is generally unique amongst an organization or a project.
   * For example,
   * all core Maven artifacts do (well, should) live under the groupId
   * org.apache.maven. Group ID's do not necessarily use the dot notation, for
   * example, the junit project. Note that the dot-notated groupId does not have
   * to correspond to the package structure that the project contains. It is,
   * however, a good practice to follow. When stored within a repository, the
   * group acts much like the Java packaging structure does in an operating
   * system. The dots are replaced by OS specific directory separators (such as
   * '/' in Unix) which becomes a relative directory structure from the base
   * repository. In the example given, the org.codehaus.mojo group lives within
   * the directory $M2_REPO/org/codehaus/mojo.
   * @default "org.acme"
   * @stability experimental
   * @featured true
   */
  readonly groupId?: string;
  /**
   * The artifactId is generally the name that the project is known by.
   * Although
   * the groupId is important, people within the group will rarely mention the
   * groupId in discussion (they are often all be the same ID, such as the
   * MojoHaus project groupId: org.codehaus.mojo). It, along with the groupId,
   * creates a key that separates this project from every other project in the
   * world (at least, it should :) ). Along with the groupId, the artifactId
   * fully defines the artifact's living quarters within the repository. In the
   * case of the above project, my-project lives in
   * $M2_REPO/org/codehaus/mojo/my-project.
   * @default "my-app"
   * @stability experimental
   * @featured true
   */
  readonly artifactId?: string;
  /**
   * Enable VSCode integration.
   * Enabled by default for root projects. Disabled for non-root projects.
   * @default true
   * @stability experimental
   */
  readonly vscode?: boolean;
  /**
   * Auto-close stale issues and pull requests.
   * To disable set `stale` to `false`.
   * @default - see defaults in `StaleOptions`
   * @stability experimental
   */
  readonly staleOptions?: github.StaleOptions;
  /**
   * Auto-close of stale issues and pull request.
   * See `staleOptions` for options.
   * @default false
   * @stability experimental
   */
  readonly stale?: boolean;
  /**
   * The README setup.
   * @default - { filename: 'README.md', contents: '# replace this' }
   * @stability experimental
   */
  readonly readme?: SampleReadmeProps;
  /**
   * The name of a secret which includes a GitHub Personal Access Token to be used by projen workflows.
   * This token needs to have the `repo`, `workflows`
   * and `packages` scope.
   * @default "PROJEN_GITHUB_TOKEN"
   * @deprecated use `projenCredentials`
   * @stability deprecated
   */
  readonly projenTokenSecret?: string;
  /**
   * Choose a method of providing GitHub API access for projen workflows.
   * @default - use a personal access token named PROJEN_GITHUB_TOKEN
   * @stability experimental
   */
  readonly projenCredentials?: github.GithubCredentials;
  /**
   * Which type of project this is (library/app).
   * @default ProjectType.UNKNOWN
   * @deprecated no longer supported at the base project level
   * @stability deprecated
   */
  readonly projectType?: ProjectType;
  /**
   * Options for mergify.
   * @default - default options
   * @deprecated use `githubOptions.mergifyOptions` instead
   * @stability deprecated
   */
  readonly mergifyOptions?: github.MergifyOptions;
  /**
   * Whether mergify should be enabled on this repository or not.
   * @default true
   * @deprecated use `githubOptions.mergify` instead
   * @stability deprecated
   */
  readonly mergify?: boolean;
  /**
   * Add a Gitpod development environment.
   * @default false
   * @stability experimental
   */
  readonly gitpod?: boolean;
  /**
   * Options for GitHub integration.
   * @default - see GitHubOptions
   * @stability experimental
   */
  readonly githubOptions?: github.GitHubOptions;
  /**
   * Enable GitHub integration.
   * Enabled by default for root projects. Disabled for non-root projects.
   * @default true
   * @stability experimental
   */
  readonly github?: boolean;
  /**
   * Add a VSCode development environment (used for GitHub Codespaces).
   * @default false
   * @stability experimental
   */
  readonly devContainer?: boolean;
  /**
   * Add a `clobber` task which resets the repo to origin.
   * @default - true, but false for subprojects
   * @stability experimental
   */
  readonly clobber?: boolean;
  /**
   * Configure options for automatic merging on GitHub.
   * Has no effect if
   * `github.mergify` or `autoMerge` is set to false.
   * @default - see defaults in `AutoMergeOptions`
   * @stability experimental
   */
  readonly autoMergeOptions?: github.AutoMergeOptions;
  /**
   * Enable automatic merging on GitHub.
   * Has no effect if `github.mergify`
   * is set to false.
   * @default true
   * @stability experimental
   */
  readonly autoMerge?: boolean;
  /**
   * Enable and configure the 'auto approve' workflow.
   * @default - auto approve is disabled
   * @stability experimental
   */
  readonly autoApproveOptions?: github.AutoApproveOptions;
  /**
   * Options for renovatebot.
   * @default - default options
   * @stability experimental
   */
  readonly renovatebotOptions?: RenovatebotOptions;
  /**
   * Use renovatebot to handle dependency upgrades.
   * @default false
   * @stability experimental
   */
  readonly renovatebot?: boolean;
  /**
   * Options for .projenrc.json.
   * @default - default options
   * @stability experimental
   */
  readonly projenrcJsonOptions?: ProjenrcJsonOptions;
  /**
   * Generate (once) .projenrc.json (in JSON). Set to `false` in order to disable .projenrc.json generation.
   * @default false
   * @stability experimental
   */
  readonly projenrcJson?: boolean;
  /**
   * The shell command to use in order to run the projen CLI.
   * Can be used to customize in special environments.
   * @default "npx projen"
   * @stability experimental
   */
  readonly projenCommand?: string;
  /**
   * The parent project, if this project is part of a bigger project.
   * @stability experimental
   */
  readonly parent?: Project;
  /**
   * The root directory of the project.
   * Relative to this directory, all files are synthesized.
   *
   * If this project has a parent, this directory is relative to the parent
   * directory and it cannot be the same as the parent or any of it's other
   * sub-projects.
   * @default "."
   * @stability experimental
   */
  readonly outdir?: string;
  /**
   * Configure logging options such as verbosity.
   * @default {}
   * @stability experimental
   */
  readonly logging?: LoggerOptions;
  /**
   * Configuration options for git.
   * @stability experimental
   */
  readonly gitOptions?: GitOptions;
  /**
   * Configuration options for .gitignore file.
   * @stability experimental
   */
  readonly gitIgnoreOptions?: IgnoreFileOptions;
  /**
   * Whether to commit the managed files by default.
   * @default true
   * @stability experimental
   */
  readonly commitGenerated?: boolean;
  /**
   * This is the name of your project.
   * @default $BASEDIR
   * @stability experimental
   * @featured true
   */
  readonly name?: string;
}
