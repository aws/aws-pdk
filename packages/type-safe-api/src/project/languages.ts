/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
/**
 * Supported languages for runtimes and infrastructure
 */
export enum Language {
  TYPESCRIPT = "typescript",
  PYTHON = "python",
  JAVA = "java",
}

/**
 * Versions of node
 */
export enum NodeVersion {
  NODE_18 = "NODE_18",
  NODE_20 = "NODE_20",
}

/**
 * Versions of java
 */
export enum JavaVersion {
  JAVA_8 = "JAVA_8",
  JAVA_11 = "JAVA_11",
  JAVA_17 = "JAVA_17",
  JAVA_21 = "JAVA_21",
}

/**
 * Versions of python
 */
export enum PythonVersion {
  PYTHON_3_11 = "PYTHON_3_11",
  PYTHON_3_12 = "PYTHON_3_12",
}

/**
 * Supported libraries for code generation
 */
export enum Library {
  TYPESCRIPT_REACT_QUERY_HOOKS = "typescript-react-query-hooks",
}

export enum WebSocketLibrary {
  TYPESCRIPT_WEBSOCKET_CLIENT = "typescript-websocket-client",
  TYPESCRIPT_WEBSOCKET_HOOKS = "typescript-websocket-hooks",
}

/**
 * Formats for documentation generation
 */
export enum DocumentationFormat {
  /**
   * HTML Documentation generated by redoc
   * @see https://github.com/Redocly/redoc
   */
  HTML_REDOC = "html_redoc",
  /**
   * OpenAPI Generator 'html2' documentation
   * @see https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/html2.md
   */
  HTML2 = "html2",
  /**
   * OpenAPI Generator 'markdown' documentation
   * @see https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/markdown.md
   */
  MARKDOWN = "markdown",
  /**
   * OpenAPI Generator 'plantuml' documentation
   * @see https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/plantuml.md
   */
  PLANTUML = "plantuml",
}
