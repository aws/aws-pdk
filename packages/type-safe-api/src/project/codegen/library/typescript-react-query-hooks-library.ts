/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { TypeScriptJsxMode } from "projen/lib/javascript";
import {
  GeneratedTypescriptLibraryProject,
  GeneratedTypescriptLibraryProjectOptions,
} from "./generated-typescript-library-project";
import { Library } from "../../languages";
import { OpenApiGeneratorHandlebarsIgnoreFile } from "../components/open-api-generator-handlebars-ignore-file";
import { GenerationOptions } from "../components/utils";

/**
 * Configuration for the generated typescript client project
 */
export interface GeneratedTypescriptReactQueryHooksProjectOptions
  extends GeneratedTypescriptLibraryProjectOptions {}

/**
 * Typescript project containing generated react-query hooks
 */
export class TypescriptReactQueryHooksLibrary extends GeneratedTypescriptLibraryProject {
  constructor(options: GeneratedTypescriptReactQueryHooksProjectOptions) {
    super({
      ...options,
      tsconfig: {
        compilerOptions: {
          jsx: TypeScriptJsxMode.REACT,
        },
      },
    });

    // Add dependencies on react-query and react
    this.addDeps("@tanstack/react-query@^4"); // Pin at 4 for now - requires generated code updates to upgrade to 5
    this.addDevDeps("react", "@types/react");
    this.addPeerDeps("react");

    // Ignore the hooks, since they're generated with the handlebars templating engine
    const hooksPattern = "**/*Hooks.ts";
    this.openapiGeneratorIgnore.addPatterns(hooksPattern);

    // The hooks are generated using the handlebars templating engine, so we include a handlebars ignore file
    const handlebarsIgnore = new OpenApiGeneratorHandlebarsIgnoreFile(this);
    handlebarsIgnore.addPatterns(`!${hooksPattern}`);
  }

  public buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "typescript-fetch",
      specPath: this.options.specPath,
      generatorDirectory: Library.TYPESCRIPT_REACT_QUERY_HOOKS,
      additionalProperties: {
        npmName: this.package.packageName,
        typescriptThreePlus: "true",
        useSingleParameter: "true",
        supportsES6: "true",
      },
      srcDir: this.srcdir,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
    };
  }
}
