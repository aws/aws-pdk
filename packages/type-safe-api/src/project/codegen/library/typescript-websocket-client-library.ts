/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import {
  GeneratedTypescriptLibraryProject,
  GeneratedTypescriptLibraryProjectOptions,
} from "./generated-typescript-library-project";
import { WebSocketLibrary } from "../../languages";
import { GenerationOptions } from "../components/utils";

/**
 * Configuration for the generated typescript websocket client project
 */
export interface TypescriptWebsocketClientLibraryOptions
  extends GeneratedTypescriptLibraryProjectOptions {}

/**
 * Typescript project containing a generated websocket client
 */
export class TypescriptWebsocketClientLibrary extends GeneratedTypescriptLibraryProject {
  constructor(options: TypescriptWebsocketClientLibraryOptions) {
    super(options);

    this.addDeps(
      // Browser + node compatible websockets
      "ws@^8",
      "isomorphic-ws@^5",
      // For sigv4 signing
      "@aws-crypto/sha256-js",
      "@aws-sdk/signature-v4",
      "@aws-sdk/protocol-http",
      "@aws-sdk/types",
      "uuid@^9"
    );
    this.addDeps("@types/ws@^8", "@types/uuid@^9");

    this.openapiGeneratorIgnore.addPatterns(
      // Skip generating http clients
      `${this.srcdir}/apis/*`,
      `${this.srcdir}/apis/**/*`
    );
  }

  public buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "typescript-fetch",
      specPath: this.options.specPath,
      generatorDirectory: WebSocketLibrary.TYPESCRIPT_WEBSOCKET_CLIENT,
      additionalProperties: {
        npmName: this.package.packageName,
        typescriptThreePlus: "true",
        useSingleParameter: "true",
        supportsES6: "true",
      },
      srcDir: this.srcdir,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
    };
  }
}
