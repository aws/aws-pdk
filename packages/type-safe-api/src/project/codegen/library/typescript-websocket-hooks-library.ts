/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { TypeScriptJsxMode } from "projen/lib/javascript";
import {
  GeneratedTypescriptLibraryProject,
  GeneratedTypescriptLibraryProjectOptions,
} from "./generated-typescript-library-project";
import { WebSocketLibrary } from "../../languages";
import { OpenApiGeneratorIgnoreFile } from "../components/open-api-generator-ignore-file";
import { GenerationOptions } from "../components/utils";

/**
 * Configuration for the generated typescript websocket hooks project
 */
export interface TypescriptWebsocketHooksLibraryOptions
  extends GeneratedTypescriptLibraryProjectOptions {
  readonly clientPackageName: string;
}

/**
 * Typescript project containing generated websocket hooks
 */
export class TypescriptWebsocketHooksLibrary extends GeneratedTypescriptLibraryProject {
  protected readonly options: TypescriptWebsocketHooksLibraryOptions;

  constructor(options: TypescriptWebsocketHooksLibraryOptions) {
    super({
      ...options,
      tsconfig: {
        compilerOptions: {
          jsx: TypeScriptJsxMode.REACT,
        },
      },
    });
    this.options = options;

    this.addDevDeps("react", "@types/react");
    this.addPeerDeps("react");

    this.openapiGeneratorIgnore.addPatterns(
      // Ignore all but relevant hooks files
      ...OpenApiGeneratorIgnoreFile.ALL_FILES_PATTERNS,
      `!${this.srcdir}/index.ts`,
      `!${this.srcdir}/hooks/*`,
      `!${this.srcdir}/hooks/**/*`
    );
  }

  public buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "typescript-fetch",
      specPath: this.options.specPath,
      generatorDirectory: WebSocketLibrary.TYPESCRIPT_WEBSOCKET_HOOKS,
      additionalProperties: {
        npmName: this.package.packageName,
        typescriptThreePlus: "true",
        useSingleParameter: "true",
        supportsES6: "true",
      },
      srcDir: this.srcdir,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
      extraVendorExtensions: {
        "x-websocket-client-package-name": this.options.clientPackageName,
      },
    };
  }
}
