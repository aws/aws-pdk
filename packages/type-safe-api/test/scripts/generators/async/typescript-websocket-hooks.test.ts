/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import os from "os";
import * as path from "path";
import { exec } from "projen/lib/util";
import { TypescriptWebsocketHooksLibrary } from "../../../../src/project/codegen/library/typescript-websocket-hooks-library";
import { withTmpDirSnapshot } from "../../../project/snapshot-utils";

describe("Typescript Async Hooks Code Generation Script Unit Tests", () => {
  it.each(["single.yaml"])("Generates With %s", (spec) => {
    const specPath = path.resolve(
      __dirname,
      `../../../resources/specs/async/${spec}`
    );

    expect(
      withTmpDirSnapshot(os.tmpdir(), (outdir) => {
        exec(`cp ${specPath} ${outdir}/spec.yaml`, {
          cwd: path.resolve(__dirname),
        });
        const project = new TypescriptWebsocketHooksLibrary({
          name: "test",
          defaultReleaseBranch: "main",
          outdir,
          specPath: "spec.yaml",
          clientPackageName: "ts-ws-client",
        });
        // Synth the project so that the generate command honours the .openapi-generator-ignore-handlebars file
        project.synth();
        exec(
          project.tasks.tryFind("create-openapitools.json")!.steps[0].exec!,
          { cwd: outdir }
        );
        exec(
          `${path.resolve(
            __dirname,
            "../../../../scripts/type-safe-api/generators/generate"
          )} ${project.buildGenerateCommandArgs()}`,
          {
            cwd: outdir,
          }
        );
        return {
          excludeGlobs: [
            ...(project as any).openapiGeneratorIgnore._patterns,
            ".projen/*",
            "spec.yaml",
          ],
        };
      })
    ).toMatchSnapshot();
  });
});
