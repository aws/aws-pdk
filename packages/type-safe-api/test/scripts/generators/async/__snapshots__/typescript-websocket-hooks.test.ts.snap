// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typescript Async Hooks Code Generation Script Unit Tests Generates With single.yaml 1`] = `{}`;
